synapse/lib/lmdb.py

```diff
@@ -0,0 +1,443 @@
+import struct
+import itertools
+
+import lmdb  # used for type resolution
+
+import synapse.lib.msgpack as s_msgpack
+
+STOR_FLAG_NOINDEX = 0x0001      # there is no byprop index for this prop
+STOR_FLAG_MULTIVAL = 0x0002     # this is a multi-value prop
+STOR_FLAG_DEFVAL = 0x0004       # Only set this if it doesn't exist
+
+class Seqn:
+    '''
+    An append optimized sequence of byte blobs.
+
+    Args:
+        lenv (lmdb.Environment): The LMDB Environment.
+        name (str): The name of the sequence.
+    '''
+    def __init__(self, lenv, name):
+
+        self.lenv = lenv
+        self.db = lenv.open_db(name)
+
+        with lenv.begin() as xact:
+
+            indx = self.nextindx(xact)
+            self.indx = itertools.count(indx)
+
+    def save(self, xact, items):
+        '''
+        Save a series of items to a sequence.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB write transaction.
+            items (tuple): The series of items to save into the sequence.
+
+        Returns:
+            None
+        '''
+        rows = []
+        for item in items:
+            byts = s_msgpack.en(item)
+            lkey = struct.pack('>Q', next(self.indx))
+            rows.append((lkey, byts))
+
+        with xact.cursor(db=self.db) as curs:
+            curs.putmulti(rows, append=True)
+
+    def nextindx(self, xact):
+        '''
+        Determine the next insert offset according to storage.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB transaction.
+
+        Returns:
+            int: The next insert offset.
+        '''
+        indx = 0
+        with xact.cursor(db=self.db) as curs:
+            if curs.last():
+                indx = struct.unpack('>Q', curs.key())[0] + 1
+        return indx
+
+    def iter(self, xact, offs):
+        '''
+        Iterate over items in a sequence from a given offset.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB transaction.
+            offs (int): The offset to begin iterating from.
+
+        Yields:
+            (indx, valu): The index and valu of the item.
+        '''
+        lkey = struct.pack('>Q', offs)
```

Comment:
Interestingly, in my latest benchmark, offs.to_bytes(8, 'little') was 6.5% faster...